source/MaterialXRuntime/Private/PvtAttribute.cpp

```diff
@@ -26,111 +28,73 @@ PvtAttribute::PvtAttribute(const RtToken& name, const RtToken& type, uint32_t fl
 }
 
 
-RT_DEFINE_RUNTIME_OBJECT(PvtOutput, RtObjType::OUTPUT, "PvtOutput")
+RT_DEFINE_RUNTIME_OBJECT(PvtInput, RtObjType::INPUT, "PvtInput")
 
-PvtOutput::PvtOutput(const RtToken& name, const RtToken& type, uint32_t flags, PvtPrim* parent) :
+PvtInput::PvtInput(const RtToken& name, const RtToken& type, uint32_t flags, PvtPrim* parent) :
     PvtAttribute(name, type, flags, parent)
 {
-    setTypeBit<PvtOutput>();
+    setTypeBit<PvtInput>();
 }
 
-bool PvtOutput::isConnectable(const PvtInput* other) const
+bool PvtInput::isConnectable(const PvtOutput* output) const
 {
     // We cannot connect to ourselves.
-    if (_parent == other->_parent)
+    if (_parent == output->_parent)
     {
         return false;
     }
-    // We must have matching types.
-    if (getType() != other->getType())
-    {
-        // TODO: Check if the data types are compatible/castable
-        // instead of strictly an exact match.
-        return false;
-    }
-    // If this is a nodegraph socket being connected to a uniform make sure
-    // the corresponding input on the outer nodegraph is also uniform.
-    // 
-    //  TODO: Enabled this check when the uniform flag has been implemented properly
-    /*
-    if (isSocket() && other->isUniform())
-    {
-        PvtPrim* graphPrim = _parent->getParent();
-        if (graphPrim)
-        {
-            PvtInput* graphInput = graphPrim->getInput(getName());
-            if (graphInput && !graphInput->isUniform())
-            {
-                return false;
-            }
-        }
-    }
-    */
-    return true;
+
+    // Use the connectable API of this prim to validate the connection.
+    RtConnectableApi* connectableApi = RtConnectableApi::get(_parent->prim());
+    return connectableApi && connectableApi->acceptConnection(hnd(), output->hnd());
 }
 
-void PvtOutput::connect(PvtInput* input)
+void PvtInput::connect(PvtOutput* output)
 {
-    // Check if the connection exists already.
-    if (input->_connection == hnd())
+    // Check if this connection exists already.
+    if (_connection == output->hnd())
     {
         return;
     }
 
-    // Validate the connection.
-    if (input->isConnected())
+    // Check if another connection exists already.
+    if (isConnected())
     {
-        throw ExceptionRuntimeError("Input '" + input->getPath().asString() + "' is already connected");
+        throw ExceptionRuntimeError("'" + getPath().asString() + "' is already connected");
     }
-    if (!isConnectable(input))
+
+    // Use the connectable API of this prim to validate the connection.
+    RtConnectableApi* connectableApi = RtConnectableApi::get(_parent->prim());
```

Comment:
Use the connectable API to validate the connection.

Reply:
Just a suggestion, but should you use isConnectable here to reduce code duplication?

Reply:
Good point! I will fix this.